src/UserAgentMiddleware.php

```diff
@@ -0,0 +1,326 @@
+<?php
+
+namespace Aws;
+
+use Aws\EndpointDiscovery\Configuration;
+use Closure;
+use Psr\Http\Message\RequestInterface;
+
+class UserAgentMiddleware
+{
+    const AGENT_VERSION = 2.1;
+    /** @var callable  */
+    private $nextHandler;
+    /** @var array */
+    private $args;
+    /** @var MetricsBuilder */
+    private $metricsBuilder;
+
+    /**
+     * Returns a middleware wrapper function.
+     *
+     * @param array $args
+     *
+     * @return Closure
+     */
+    public static function wrap(
+        array $args
+    ) : Closure
+    {
+        return function (callable $handler) use ($args) {
+            return new self($handler, $args);
+        };
+    }
+
+    /**
+     * @param callable $nextHandler
+     * @param array $args
+     */
+    public function __construct(callable $nextHandler, array $args=[])
+    {
+        $this->nextHandler = $nextHandler;
+        $this->args = $args;
+    }
+
+    /**
+     * When invoked, its injects the user agent header into the
+     * request headers.
+     * @param CommandInterface $command
+     * @param RequestInterface $request
+     *
+     * @return mixed
+     */
+    public function __invoke(CommandInterface $command, RequestInterface $request)
+    {
+        $handler = $this->nextHandler;
+        $this->metricsBuilder = MetricsBuilder::fromCommand($command);
+        $request = $this->requestWithUserAgentHeader($request);
+
+        return $handler($command, $request);
+    }
+
+    /**
+     * Builds the user agent header value, and injects it into the request
+     * headers. Then, it returns the mutated request.
+     *
+     * @param RequestInterface $request
+     *
+     * @return RequestInterface
+     */
+    private function requestWithUserAgentHeader(RequestInterface $request): RequestInterface
+    {
+        $uaAppend = $this->args['ua_append'] ?? [];
+        $userAgentValue = array_merge(
+            $this->buildUserAgentValue(),
+            $uaAppend
+        );
+        // It includes the user agent values just for the User-Agent header.
+        // The reason is that the SEP does not mention appending the
+        // metrics into the X-Amz-User-Agent header.
+        return $request->withHeader(
+            'X-Amz-User-Agent',
+            implode(' ', array_merge(
+                $uaAppend,
+                $request->getHeader('X-Amz-User-Agent')
+            ))
+        )->withHeader(
+            'User-Agent',
+            implode(' ', array_merge(
+                $userAgentValue,
+                $request->getHeader('User-Agent')
+            ))
+        );
+    }
+
+    /**
+     * Builds the different user agent values.
+     *
+     * @return array
+     */
+    private function buildUserAgentValue(): array
+    {
+        static $fnList = [
```

Comment:
suggestion: consider moving to a static property if it could possibly be used by anything else

Reply:
Comment still applies